first attempt for issue #326

dafnapension opened the pull request.

yoavkatz reviewed (commented):
Why do we need a new Operator, can't we extend the existing MapIstanceValues operator, to just have the "process_every_value" field?

yoavkatz commented on src/unitxt/operators.py:
```diff
@@ -100,6 +100,72 @@ def process(self, instance: Dict[str, Any], stream_name: str = None) -> Dict[str
+            When process_every_value=True: apply also if the content of 'f' is a list, one of whose members is 'v'.
+            In such a case, replace just member 'v' in the list, by 'm[v]', leaving the other members of the list as they are
+            (potentially, to be replaced by other values of 'm').
+    """
```
I think the most important point is to add examples.

yoavkatz commented on src/unitxt/operators.py:
```diff
@@ -100,6 +100,72 @@ def process(self, instance: Dict[str, Any], stream_name: str = None) -> Dict[str
+        strict (bool): If True, the mapping is applied strictly. That means if a value
+            does not exist in the mapper, it will raise a KeyError. If False, values
+            that are not present in the mapper are kept as they are.
+        process_every_value (bool): Let 'm' denote the mapper associated with field 'f', and let 'v' be a key in 'm'.
```
Better to avoid formal notations and simplify for the user.  .  All this means, is that if prices_every_value =True is that, each field mist be a  list, and the mapping is done to individual elements in the list.

dafnapension commented:
Hi @yoavkatz ,
You are absolutely right in all your comments. 
Such an elegant solution of just adding a new parameter did not occur to me at all..
Simplified now implementation and documentation. Added examples.

yoavkatz commented on src/unitxt/operators.py:
```diff
@@ -71,32 +71,64 @@ class MapInstanceValues(StreamInstanceOperator):
+            is to be applied to their individual elements. If False, mapping is only applied to a field
+            containing a single value.
+
+    Examples: MapInstanceValues with mappers={"a": {"1": "hi", "2": "bye"}}, and process_every_value = True,
```
@dafnapension @elronbandel - I think we should agree on the example format for all operators.  I think we should aim that each example will be self contained, and one could cut and past it.  Fro example:

Replace '1' with 'hi' and '2' with 'bye' in field 'a' of all streams:
`MapInstanceValues(mappers={"a": {"1": "hi", "2": "bye"}})`
So {"a":1 , "b": 2} becomes "a":"hi" , "b": 2} 

Assuming 'a' is field with a list of values, replace  all 1s with 'hi' and all 2s with 'bye'
`MapInstanceValues(mappers={"a": {"1": "hi", "2": "bye"}}, process_every_element=True`
 So {"a": [1, 2], "b": 2} becomes {"a": ["hi", "bye"], "b": 2}.

 Ensure all values are mapped in the instance
 `MapInstanceValues(mappers={"a": {"1": "hi", "2": "bye"}}, strict=True)`
 So if the input is  {"a":3 , "b": 2} it will throw an exception, since the mapping of 3 in not in the mapper.

dafnapension replied:
I accept.

yoavkatz commented on tests/test_operators.py:
```diff
@@ -45,23 +46,86 @@ def compare_streams(self, all, expected_all):
         inputs = [
-            {"a": 1, "b": 2},
-            {"a": 2, "b": 3},
+            {"a": "1", "b": "2"},
```
I thinking keeping the input  as numbers is a more realistic example.

dafnapension replied:
Hi @yoavkatz ,
We have a hidden problem here: the mappers are defined as  str->str dictionaries, and when processing, the input values are first transformed to strings, then looked up at the dictionary, and the output is str again. Not back to int.  So any mapped value is always a str.  Could this make a problem?

yoavkatz replied:
I think the use of string internally is fine.  but people use it to map inputs that are not string, so we should keep this example.

Usually from int to string - as above.

yoavkatz commented on tests/test_operators.py:
```diff
@@ -45,23 +46,86 @@ def compare_streams(self, all, expected_all):
         )
 
+        # process_every_value=True would not accept non-list inputs
+        with self.assertRaises(ValueError):
```
Please check exception with the new `check_operator_exception` that also checks the exception text.

    exception_text = "Required filter field ('c') in FilterByValues is not found in {'a': 1, 'b': 2}"
        check_operator_exception(
            operator=FilterByValues(required_values={"c": "5"}),
            inputs=inputs,
            exception_text=exception_text,
            tester=self,
        )

yoavkatz commented on tests/test_operators.py:
```diff
@@ -45,23 +46,86 @@ def compare_streams(self, all, expected_all):
+            operator = MapInstanceValues(mappers=mappers)
+            operator.process(instance={"a": "3", "b": "4"})
+
+        inputs_p_e_v = [
```
Why name is p_e_v?

dafnapension replied:
to be used with example case with process_every_value  is True.  Will extend to full name..

yoavkatz reviewed (changes requested):
Hi @dafnapension . Please see a few comments below.

yoavkatz commented on src/unitxt/operators.py:
```diff
@@ -71,32 +71,64 @@ class MapInstanceValues(StreamInstanceOperator):
+
     """
 
     mappers: Dict[str, Dict[str, str]]
```
I think this should be Dict[str, Dict[any, str]]

yoavkatz commented on src/unitxt/operators.py:
```diff
@@ -75,13 +75,19 @@ class MapInstanceValues(StreamInstanceOperator):
+    Examples:
 
+        Replace '1' with 'hi' and '2' with 'bye' in field 'a' in all instances of all streams:
+        MapInstanceValues(mappers={"a": {"1": "hi", "2": "bye"}}) will modify, for example,
```
```suggestion
        MapInstanceValues(mappers={"a": {"1": "hi", "2": "bye"}}) 
```

yoavkatz commented on src/unitxt/operators.py:
```diff
@@ -75,13 +75,19 @@ class MapInstanceValues(StreamInstanceOperator):
 
+        Replace '1' with 'hi' and '2' with 'bye' in field 'a' in all instances of all streams:
+        MapInstanceValues(mappers={"a": {"1": "hi", "2": "bye"}}) will modify, for example,
+        instance {"a":"1", "b": 2} to become {"a":"hi", "b": 2}.
```
```suggestion
        instance {"a":"1", "b": 2} to become {"a":"hi", "b": 2}.
        
```

yoavkatz commented on src/unitxt/operators.py:
```diff
@@ -75,13 +75,19 @@ class MapInstanceValues(StreamInstanceOperator):
+        instance {"a":"1", "b": 2} to become {"a":"hi", "b": 2}.
+        Assuming field 'a' is a list of values, potentially including "1"-s and "2"-s, replacing each
+        such "1" with "hi" and "2" -- with "bye" can be achieved by employing
+        MapInstanceValues(mappers={"a": {"1": "hi", "2": "bye"}}, process_every_element=True). This will modify,
```
```suggestion
        MapInstanceValues(mappers={"a": {"1": "hi", "2": "bye"}}, process_every_element=True)
```

yoavkatz commented on src/unitxt/operators.py:
```diff
@@ -75,13 +75,19 @@ class MapInstanceValues(StreamInstanceOperator):
+        for example, instance {"a": ["1", "2"], "b": 2} to become {"a": ["hi", "bye"], "b": 2}.
+
+        To ensure that all values of field 'a' are mapped in every instance, use
+        MapInstanceValues(mappers={"a": {"1": "hi", "2": "bye"}}, strict=True). Here, input instance
```
```suggestion
        MapInstanceValues(mappers={"a": {"1": "hi", "2": "bye"}}, strict=True
```

dafnapension commented:
Thanks, @yoavkatz , 
I accept your good comments, and implemented.
